File: tarjan.py
"""
Tarjan's strongly connected components algorithm is an algorithm in graph theory for finding the strongly 
connected components (SCCs) of a directed graph. It runs in linear time, matching the time bound for alternative 
methods including Kosaraju's algorithm and the path-based strong component algorithm. The algorithm is named for its 
inventor, Robert Tarjan.
"""

import sys
sys.setrecursionlimit(10 ** 5)

n, m = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(n + 1)]

for _ in range(m):
    a, b = map(int, sys.stdin.readline().split())
    graph[a].append(b)

scc = []
stack = []
finish = [0] * (n + 1)
visit = [0] * (n + 1)
idx = 1


def dfs(cur):
    global idx
    visit[cur] = idx
    low = idx
    idx += 1
    stack.append(cur)

    for nxt in graph[cur]:
        if not visit[nxt]:
            low = min(low, dfs(nxt))
        elif not finish[nxt]:
            low = min(low, visit[nxt])

    if low == visit[cur]:
        res = []
        while stack:
            top = stack.pop()
            finish[top] = 1
            res.append(top)
            if cur == top:
                break
        scc.append(res)

    return low


for i in range(1, n + 1):
    if not visit[i]:
        dfs(i)

File: tarjan.cpp
// BOJ 2150 Strongly Connected Component

#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

int dfs(int cur, int &idx, vector<int> &v, vector<int> &vst, vector<bool> &finish, vector<vector<int>> &graph, vector<vector<int>> &scc) {
    v.push_back(cur);
    vst[cur] = idx;
    int low = idx++;

    for (int nxt : graph[cur]) {
        if (!vst[nxt])
            low = min(low, dfs(nxt, idx, v, vst, finish, graph, scc));
        else if (!finish[nxt])
            low = min(low, vst[nxt]);
    }

    if (low != vst[cur])
        return low;

    vector<int> w;
    while (1) {
        int top = v.bk;
        v.pop_back();

        finish[top] = true;
        w.push_back(top);

        if (cur == top)
            break;
    }

    scc.push_back(w);
    return low;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n + 1);

    while (m--) {
        int x, y;
        cin >> x >> y;

        graph[x].push_back(y);
    }

    int idx = 1;
    vector<int> v, vst(n + 1);
    vector<bool> finish(n + 1);
    vector<vector<int>> scc;

    for (int i = 1; i <= n; i++)
        if (!vst[i])
            dfs(i, idx, v, vst, finish, graph, scc);

    for (auto &w : scc)
        sort(w.begin(), w.end());

    sort(scc.begin(), scc.end());

    cout << scc.sz << '\n';
    for (auto &w : scc) {
        for (int x : w)
            cout << x << ' ';
        cout << -1 << '\n';
    }
}
